make the osd as large as the screen for vo=drm

sgerwk opened the pull request:
I agree that my changes can be relicensed to LGPL 2.1 or later.

wm4 commented on video/out/vo_drm.c:
```diff
@@ -320,21 +314,33 @@ static void draw_image(struct vo *vo, mp_image_t *mpi)
             mp_image_crop_rc(&src, src_rc);
+
+            w = p->dst.x1 - p->dst.x0;
+	    h = p->dst.y1 - p->dst.y0;
```
Don't use tabs for indentation.

wm4 commented:
CC @rr- (I think?)

sgerwk commented on video/out/vo_drm.c:
```diff
@@ -320,21 +314,33 @@ static void draw_image(struct vo *vo, mp_image_t *mpi)
+
+            w = p->dst.x1 - p->dst.x0;
+            h = p->dst.y1 - p->dst.y0;
+            x = (p->cur_frame->w - w) >> 1;
```
I tried to follow the same coding style of the original file, which had these variables signed and >>1 instead of /2. I agree that at least w and h should be unsigned, and dividing by two is clearer than shifting.

sgerwk commented on video/out/vo_drm.c:
```diff
@@ -320,21 +314,33 @@ static void draw_image(struct vo *vo, mp_image_t *mpi)
+
+            shift = x * BYTES_PER_PIXEL + y * p->cur_frame->stride[0];
+            temp = p->cur_frame->planes[0];
+            p->cur_frame->planes[0] += shift;
```
Setting up an avfilter with pad=..., maybe. But seems that no other vo does anything like this; I don't know if it's really doable this way.

The original vo_drm.c does shifting in the same way, by adding an offset to the pointer to an image being copied. The difference is that then p->cur_frame->planes[0] was the source of the copy, in my version is the target.

sgerwk replied:
Wait, maybe I misunderstood your comment. Did you mean something like creating a new image with image->planes[0] equal to p->cur_frame->planes[0] + shift?

xantoz replied:
Yes. I had something like that in mind (the latter).

sgerwk commented:
I created a dummy reference to the cur_crop image and then cropping it. This way, swscale on the dummy reference gives the right result. The code is also simplified this way.

Perhaps it also makes sense to move the four clear_area to mp_image.c, something like mp_image_clear_outside() and mp_image_clear_outside_rc().

xantoz commented:
seems like 8c9e9f3 fixes https://github.com/mpv-player/mpv/issues/5438

kevmitch commented:
All three commits appear to work as advertised. Importantly, the first one doesn't seem to have significant performance impact playing a 4:3 1080p on a 16:9 1080p display.

Could you put those messages in the commits themselves (wrapped to 72 columns)?

As per [contribute.md](https://github.com/mpv-player/mpv/blob/master/DOCS/contribute.md), the commit message should talk about the behaviour before the commit in past tense and use present tense for the new state of affairs.

> which can then be added the osd 

"which can then have the osd added to it"

> The last image is stored in vo->priv->last_input to be used when redrawing a freme is necessary (contol: VOCTRL_REDRAW_FRAME).

frame, control

>  For example, if drm mode 5 is "640x400" the right aspect on a 4:3 monitor is obtained by mpv 

comma after "640x400" 

so the monitorpixel aspect must be set manually? I only have a single 1:1 drm mode here so can't test.

sgerwk commented:
Fixed comments, moved them to the commit messages.

Yes, you currently have to set --monitorpixelaspect manually, otherwise 1:1 pixels are assumed.

Actually, the drm modes are suggestions, not the only resolutions that can be used. In X11, you can see the drm modes by xrandr, but then add new ones like (assuming the connector you use is LVDS):

cvt 640 400
xrandr --newmode (output of cvt, with the word "Modeline" removed)
xrandr --addmode LVDS (mode name: first string in cvt output after "Modeline")
xrandr --output LVDS --mode (mode name)

In the linux vt, the only way I found to have a new drm mode is to add it to the kernel parameters at startup, like "video=640x400R-16". Actually, --drm-mode=n is a limitation, since it constraints mpv to run only in one of the modes that are present at startup.

kevmitch commented:
Thanks. That's much better. Just a couple more tweaks

> the drm vo drawn the osd over the scaled image

the drm vo drew 

> This commit deallocates and reset to NULL

deallocates and reset**s**

kevmitch commented:
thanks.
